src/app/directives/restrict-special-char.directive.ts

```diff
@@ -0,0 +1,25 @@
+import { Directive, HostListener, Input } from '@angular/core';
+
+@Directive({
+  selector: '[appRestrictSpecialChars]'
+})
+export class RestrictSpecialCharDirective {
+  @Input() restrictedCharCodes: any[];
```

Comment:
What is the purpose of @Input() ?

Reply:
We can pass any value to `restrictedCharCodes` from any component and its value will be an array of the character codes which we want to restrict to enter in any input.

Reply:
As you said `restrictedCharCodes` is always an array then you can declare restrictedCharCodes as `@Input() restrictedCharCodes = [];`. By doing this you can remove the undefined check in if block on line 19.